src/org/ods/orchestration/usecase/BitbucketTraceabilityUseCase.groovy

```diff
@@ -0,0 +1,170 @@
+package org.ods.orchestration.usecase
+
+import org.ods.orchestration.util.Project
+import org.ods.services.BitbucketService
+import org.ods.util.IPipelineSteps
+
+import java.text.SimpleDateFormat
+
+class BitbucketTraceabilityUseCase {
+
+    private static final String CSV_FILE = "source-code-review.csv"
+    private static final int PAGE_LIMIT = 10
+
+    private final BitbucketService bitbucketService
+    private final IPipelineSteps steps
+    private final Project project
+
+    BitbucketTraceabilityUseCase(BitbucketService bitbucketService, IPipelineSteps steps, Project project) {
+        this.steps = steps
+        this.project = project
+        this.bitbucketService = bitbucketService
+    }
+
+    /**
+     * Create a CSV file that contains the following records
+     * for every merge event into the integration branch of every ODS component:
+     * @return absolutePath of the created file
+     */
+    @SuppressWarnings(['JavaIoPackageAccess'])
+    String generateSourceCodeReviewFile() {
+        def file = new File("${steps.env.WORKSPACE}/${CSV_FILE}")
+        if (file.exists()) {
+            file.delete()
+        }
+        file.createNewFile()
+
+        def token = bitbucketService.getToken()
+        List<Map> repos = getRepositories()
+        repos.each {
+            processRepo(token, it, file)
+        }
+        return file.absolutePath
+    }
+
+    private void processRepo(String token, Map repo, File file) {
+        def nextPage = true
+        def nextPageStart = 0
+        while (nextPage) {
+            def commits = bitbucketService.getCommitsForMainBranch(token, repo.repo, PAGE_LIMIT, nextPageStart)
+            if (commits.isLastPage) {
+                nextPage = false
+            } else {
+                nextPageStart = commits.nextPageStart
+            }
+            processCommits(token, repo, commits, file)
```

Comment:
If I read this right, this make one API request per page until all commits have been found, and for each commit, make another API request to find any pull requests.

This sounds like a very expensive operation - which will get more expensive over time. Did you calculate the impact on the Bitbucket Server?

Reply:
Yes, that is the process. And no, right now we don't have calculated the impact, we are making test to probe it.

Reply:
Hector, found out that we reach a limit at ~8k commits. The reason is not yet clear, since there is no error being shown. @hrcornejo @jorge-romero we should consider only including those commits that belong to the current product version. This requires keeping track of the commit SHA of the previous releases.

Reply:
Would other endpoints help us instead? E.g. going through the merged pull requests instead, and then plucking the ones we want to use?

Reply:
Ah. I think there is another endpoint that gives you "PR actions" or something like this which does show the merged commit. I would have to look it up

Reply:
@hrcornejo @michaelsauter we should definitely investigate this.